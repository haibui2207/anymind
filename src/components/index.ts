export { default as Select } from "./Select";
export { default as StaticHTML } from "./StaticHTML";
