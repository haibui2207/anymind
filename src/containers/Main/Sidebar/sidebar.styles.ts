import { createUseStyles } from "react-jss";

import { rem } from "../../../utils/jss";
import COLORS from "../../../constants/COLORS";
import BREAKPOINTS from "../../../constants/BREAKPOINTS";

export default createUseStyles({
  root: {
    width: rem(320),
    boxSizing: "border-box",
    padding: [rem(16), rem(16), rem(24)],
    backgroundColor: COLORS.bgBlack,
    height: "100%",
    transform: "translateX(0)",
    transition: "width .5s ease, padding .3s ease, transform .5s ease",
  },
  infoWrapper: { display: "flex", alignItems: "center", marginBottom: rem(32) },
  infoTitle: {
    marginTop: 0,
    marginBottom: rem(12),
    fontSize: rem(14),
    lineHeight: rem(20),
    color: COLORS.textWhite,
  },
  infoGroup: { flex: 1 },
  avatarWrapper: {
    flex: "none",
    backgroundColor: COLORS.bgWhite,
    overflow: "hidden",
    borderRadius: "50%",
    width: rem(50),
    height: rem(50),
    display: "flex",
    alignItems: "center",
    justifyContent: "center",
    marginRight: rem(12),
  },
  avatar: { width: "100%" },
  title: {
    margin: [rem(16), 0, rem(8)],
    fontSize: rem(20),
    lineHeight: rem(24),
    color: COLORS.textWhite,
  },
  listChannels: {
    margin: 0,
    padding: 0,
    listStyle: "none",
  },
  channelItem: {},
  link: {
    display: "block",
    color: COLORS.textGray,
    textDecoration: "none",
    padding: [rem(12)],
    fontSize: rem(16),
    lineHeight: rem(24),
    "&.active": {
      fontWeight: "bold",
      color: COLORS.textWhite,
    },
  },

  [`@media only screen and (max-width: ${BREAKPOINTS.maxLg}px)`]: {
    root: { width: rem(260) },
  },
  [`@media only screen and (max-width: ${BREAKPOINTS.maxMd}px)`]: {
    root: { width: rem(240), padding: [rem(16), rem(8), rem(24)] },
    infoWrapper: { marginBottom: rem(24) },
    infoTitle: { fontSize: rem(12), lineHeight: rem(16) },
    avatarWrapper: { width: rem(40), height: rem(40), marginRight: rem(8) },
    title: { fontSize: rem(16), lineHeight: rem(20) },
    link: { padding: [rem(8)], fontSize: rem(12), lineHeight: rem(16) },
  },
  [`@media only screen and (max-width: ${BREAKPOINTS.maxSm}px)`]: {
    root: {
      position: "absolute",
      top: 0,
      left: 0,
      "&:not(.show)": { transform: "translateX(-100%)" },
    },
    toggleButton: {
      display: "block",
      position: "absolute",
      right: 0,
      width: rem(50),
      height: rem(50),
      transform: "translateX(100%)",
      color: COLORS.textWhite,
      cursor: "pointer",
      backgroundColor: COLORS.bgBlack,
      border: "none",
    },
  },
});
