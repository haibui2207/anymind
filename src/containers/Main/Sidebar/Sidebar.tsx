import React, { memo, useCallback, useEffect, useRef } from "react";
import classNames from "classnames";

import { Select } from "../../../components";
import { IChannel, IUser } from "../../../models";
import { useConversationContext } from "../../../providers/ConversationProvider";

import useStyles from "./sidebar.styles";

const Sidebar: React.FC<{}> = () => {
  const ref = useRef<HTMLDivElement>(null);
  const classes = useStyles();
  const {
    users,
    channels,
    activeUser,
    activeChannel,
    activeSidebar,
    setActiveUser,
    setActiveChannel,
    setActiveSidebar,
  } = useConversationContext();

  const handleChannelClick = useCallback(
    (channel: IChannel): void => {
      if (activeChannel?.id !== channel.id) {
        setActiveChannel(channel);
        setActiveSidebar(false);
        // TODO fetch latest message when channel change
      }
    },
    [activeChannel]
  );

  const handleUserSelectChange = useCallback(
    (user: IUser): void => {
      if (activeUser?.id !== user.id) {
        setActiveUser(user);
        // TODO fetch latest message when user change
      }
    },
    [activeUser]
  );

  useEffect(() => {
    const listener = (event: any): any => {
      if (!ref.current || ref.current.contains(event.target)) {
        return;
      }

      if (activeSidebar) {
        setActiveSidebar(!activeSidebar);
      }
    };

    document.addEventListener("mousedown", listener);
    document.addEventListener("touchstart", listener);

    return () => {
      document.removeEventListener("mousedown", listener);
      document.removeEventListener("touchstart", listener);
    };
  }, [ref, activeSidebar]);

  return (
    <div
      ref={ref}
      className={classNames("custom-scrollbar", classes.root, {
        show: activeSidebar,
      })}
    >
      <div className={classes.infoWrapper}>
        <div className={classes.avatarWrapper}>
          <img
            alt={activeUser?.name}
            src={activeUser?.avatar}
            className={classes.avatar}
            data-test="sidebar-active-user-avatar"
          />
        </div>
        <div className={classes.infoGroup}>
          <h4 className={classes.infoTitle}>Switch user</h4>
          <Select
            name="user"
            options={users}
            onChange={({ value }: { value: IUser }) =>
              handleUserSelectChange(value)
            }
            value={activeUser}
            getOptionLabel={(user: IUser): string => user.name}
            getOptionValue={(user: IUser): string => user.name}
          />
        </div>
      </div>
      <h4 className={classes.title}>Channels</h4>
      <ul className={classes.listChannels}>
        {channels.map((channel: IChannel) => (
          <li
            key={channel.id}
            className={classes.channelItem}
            data-test="sidebar-channel-item"
          >
            <a
              href="/#"
              type="button"
              className={classNames(classes.link, {
                active: activeChannel?.id === channel.id,
              })}
              onClick={(e) => {
                e.preventDefault();
                handleChannelClick(channel);
              }}
            >
              #&nbsp;&nbsp;{channel.name}
            </a>
          </li>
        ))}
      </ul>
    </div>
  );
};

export default memo(Sidebar);
