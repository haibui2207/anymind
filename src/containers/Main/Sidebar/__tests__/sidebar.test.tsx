import React from "react";
import { shallow, mount } from "enzyme";
import toJson from "enzyme-to-json";

import { Select } from "../../../../components";
import {
  IConversation,
  ConversationContext,
  defaultContextValue,
} from "../../../../providers/ConversationProvider";

import Sidebar from "../Sidebar";

describe("Testing Main > Sidebar", (): void => {
  let wrapper: any;
  const mockContextValue: IConversation = {
    ...defaultContextValue,
    users: [{ id: "1", name: "User 1", avatar: "" }],
    channels: [{ id: "1", name: "Channel 1" }],
  };

  beforeAll((): void => {
    wrapper = mount(
      <ConversationContext.Provider value={mockContextValue}>
        <Sidebar />
      </ConversationContext.Provider>
    );
  });

  it("should render match snapshot", (): void => {
    const wrapper = shallow(
      <ConversationContext.Provider value={mockContextValue}>
        <Sidebar />
      </ConversationContext.Provider>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("should show current active user avatar", (): void => {
    expect(
      wrapper.find('[data-test="sidebar-active-user-avatar"]')
    ).toHaveLength(1);
  });

  it("should have an Select to switch user", (): void => {
    expect(wrapper.find(Select)).toHaveLength(1);
  });

  it("should have some channels", (): void => {
    expect(wrapper.find('[data-test="sidebar-channel-item"]')).toHaveLength(
      mockContextValue.channels.length
    );
  });
});
