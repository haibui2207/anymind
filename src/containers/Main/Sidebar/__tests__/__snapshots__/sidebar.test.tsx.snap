// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Main > Sidebar should render match snapshot 1`] = `<Memo(Sidebar) />`;
